src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/Marshal.cs

```diff
@@ -891,7 +922,27 @@ public static Delegate GetDelegateForFunctionPointer(IntPtr ptr, Type t)
 
         public static TDelegate GetDelegateForFunctionPointer<TDelegate>(IntPtr ptr)
         {
-            return (TDelegate)(object)GetDelegateForFunctionPointer(ptr, typeof(TDelegate));
+            if (ptr == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(nameof(ptr));
+            }
+
+            Type t = typeof(TDelegate);
+            if (t.IsGenericType)
+            {
+                throw new ArgumentException(SR.Argument_NeedNonGenericType, nameof(TDelegate));
+            }
+
+            // COMPAT: This block of code isn't entirely correct.
+            // Users passing in typeof(MulticastDelegate) as 't' skip this check
+            // since Delegate is a base type of MulticastDelegate.
+            Type? c = t.BaseType;
+            if (c != typeof(Delegate) && c != typeof(MulticastDelegate))
+            {
+                throw new ArgumentException(SR.Arg_MustBeDelegate, nameof(TDelegate));
+            }
```

Comment:
This odd check should not be needed here at all.

The `MulticastDelegate` corner case needs to be handled on the unmanaged side. It is handled correctly for CoreCLR already.

`Marshal.GetDelegateForFunctionPointer<MulticastDelegate>(Marshal.AllocHGlobal(1))` should throw `Type must derive from Delegate.`. It does so in CoreCLR already even with this check deleted, but not in CoreRT.

Reply:
(We can deal with this separately.)

Reply:
This check is actually needed. I am updating the comment and adding a test in dotnet/runtime